### PR TITLE
[15.0] add account_block_payment

### PR DESCRIPTION
**Pull request by victorvermot**
I built a module that allows Odoo users to block the payment of invoices.
The user can set a boolean field on each invoice to make it impossible to register their payments.

### CODE REVIEW

File: account_invoice_block_payment/models/account_move.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    blocked = fields.Boolean(
+        string="Payment Blocked",
+        default=False,
```
Comment (gurneyalex) [inline account_invoice_block_payment/models/account_move.py]: this is the normal default -> redundant -> remove 
```suggestion

```

File: account_invoice_block_payment/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        help="It indicates whether the payment of the invoice is blocked.",
```
Comment (gurneyalex) [inline account_invoice_block_payment/models/account_move.py]:
```suggestion
        help="if true, the payment of the invoice is blocked.",
```

File: account_invoice_block_payment/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    blocked = fields.Boolean(
```
Comment (gurneyalex) [inline account_invoice_block_payment/models/account_move.py]: "blocked" is too generic here, I recommend using a more descriptive name. 

```suggestion
    payment_blocked = fields.Boolean(
```

TODO: rename everywhere

File: account_invoice_block_payment/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from odoo import _, fields, models
```
Comment (gurneyalex) [inline account_invoice_block_payment/models/account_move.py:4]: insert copyright and license, on all files

File: account_invoice_block_payment/models/account_move.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        help="It indicates whether the payment of the invoice is blocked.",
+    )
+
+    def action_register_payment(self):
+        error_msg = ""
+        for move in self:
+            if move.blocked:
+                error_msg += _("The payment on invoice {} is blocked.\n").format(
+                    move.name
+                )
```
Comment (gurneyalex) [inline account_invoice_block_payment/models/account_move.py]:
```suggestion
                error_msg += _("The payment on invoice {} is blocked.").format(
                    move.name
                ) + "\n"
```

File: account_invoice_block_payment/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1,7 @@
+==================================
+Block Payment of Invoices
+==================================
```
Comment (gurneyalex) [inline account_invoice_block_payment/readme/DESCRIPTION.rst:3]:
```suggestion
=========================
Block Payment of Invoices
=========================
```

File: account_invoice_block_payment/readme/DESCRIPTION.rst [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+
+This module allow the user to set invoices payment as blocked.
```
Comment (gurneyalex) [inline account_invoice_block_payment/readme/DESCRIPTION.rst]:
```suggestion
This module allows the user to block the payment of an invoice.
```

File: account_invoice_block_payment/readme/DESCRIPTION.rst [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+If the user tries to register payment for a blocked invoice, an error will be raised and the user will
+be notified on which invoice payment was blocked.
```
Comment (gurneyalex) [inline account_invoice_block_payment/readme/DESCRIPTION.rst:7]:
```suggestion
be notified on which invoice the payment was blocked.
```

File: account_invoice_block_payment/readme/USAGE.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+The user can set a flag on each invoices in order to block their payment.
```
Comment (gurneyalex) [inline account_invoice_block_payment/readme/USAGE.rst]:
```suggestion
The user can set a flag on each invoice in order to block its payment.
```

File: account_invoice_block_payment/tests/test_account_block_payment.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+class TestAccountInvoice(TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        # create an invoice with its payment unblocked
+        cls.account_move = cls.env["account.move"].create({"blocked": False})
+
+    def test_payment_blocked(self):
+        # Blocked boolean set to true rises a ValidationError
+        self.assertFalse(self.account_move.blocked)
```
Comment (gurneyalex) [inline account_invoice_block_payment/tests/test_account_block_payment.py]:
```suggestion
        self.assertFalse(self.account_move.payment_blocked)
```

File: account_invoice_block_payment/tests/test_account_block_payment.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+        self.account_move.blocked = True
```
Comment (gurneyalex) [inline account_invoice_block_payment/tests/test_account_block_payment.py]:
```suggestion
        self.account_move.payment_blocked = True
```

File: account_invoice_block_payment/tests/test_account_block_payment.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        with self.assertRaises(UserError):
+            self.account_move.action_register_payment()
+        self.account_move.blocked = False
```
Comment (gurneyalex) [inline account_invoice_block_payment/tests/test_account_block_payment.py]:
```suggestion
        self.account_move.payment_blocked = False
```

File: account_invoice_block_payment/views/account_move.xml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2016 Acsone SA/NV
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record id="view_move_form" model="ir.ui.view">
+        <field name="model">account.move</field>
+        <field name="inherit_id" ref="account.view_move_form" />
+        <field name="arch" type="xml">
+            <field name="payment_reference" position="after">
+                <field name="blocked" />
```
Comment (gurneyalex) [inline account_invoice_block_payment/views/account_move.xml]:
```suggestion
                <field name="payment_blocked" />
```

Comment (gurneyalex) [reply account_invoice_block_payment/views/account_move.xml]: This field is always visible: is this the intent (I'm not sure it makes sense to have the field visible if the invoice is already paid). Check the visibility rules for the button to register payment, I think the visibility rules for the field should be the same:

https://github.com/odoo/odoo/blob/e845e458548534acbac55bd355cdcdd295b4ae2f/addons/account/views/account_move_views.xml#L632-L637

With this in mind, I think the button which calls action_register_payment should be hidden if payment is blocked. Can you check this (and the button is also available in the tree view)

File: account_invoice_block_payment/models/account_move.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    )
+
+    def action_register_payment(self):
+        error_msg = ""
+        for move in self:
+            if move.payment_blocked:
+                error_msg += (
+                    _("The payment on invoice {} is blocked.").format(move.name) + "\n"
+                )
+        if len(error_msg):
```
Comment (gurneyalex) [inline account_invoice_block_payment/models/account_move.py]:
```suggestion
        if error_msg:
```

In Python a character string which is not empty is considered as True

File: account_invoice_block_payment/readme/DESCRIPTION.rst (hunk 7 above)
Comment (gurneyalex) [inline account_invoice_block_payment/readme/DESCRIPTION.rst]:
```suggestion
This module allows the user to set invoices payment as blocked.
```

**Review by gurneyalex — APPROVED** [review]
2 small things. 

It does feel a bit strange to have this applied on customer invoices and supplier refunds, but why not...

**Review by fclementic2c — APPROVED** [review]
Functional test Ok

**Pull request merged**